pkg/query/args.go

```diff
@@ -102,6 +104,186 @@ type Args struct {
 	outputs []io.Writer
 }
 
+// ArgsError provides a more detailed error description for invalid query args
+type ArgsError struct {
+	Field   string // Field: the string describing which field led to the error. It MUST match the json definition for a field. Example: condition
+	Type    string // Type: the type of the error. Example: *types.ParseError
+	Message string // Message: a human-readable, UI friendly description of the error. Example: Condition parsing failed
+	err     error  // Details: a specific error describing which part of the arguments is incorrect
+}
+
+func newArgsError(field string, msg string, err error) *ArgsError {
+	args := &ArgsError{
+		Field:   field,
+		Message: msg,
+		err:     err,
+	}
+	if err != nil {
+		args.Type = fmt.Sprintf("%T", err)
+
+		// make sure the type of the wrapped error is found out
+		e := errors.Unwrap(err)
+		if e != nil {
+			args.Type = fmt.Sprintf("%T", e)
+		}
+	}
+	return args
+}
+
+func (err *ArgsError) String() string {
+	return fmt.Sprintf("%s: %s: (%s: %s)", err.Field, err.Message, err.Type, err.err)
+}
+
+// Error implements the error interface
+func (err *ArgsError) Error() string {
+	return err.String()
+}
+
+// Unwrap makes the error wrappable
+func (err *ArgsError) Unwrap() error {
+	return err.err
+}
+
+// LogValue implements slog.LogValuer
+func (err *ArgsError) LogValue() slog.Value {
+	return slog.GroupValue(
+		slog.String("field", err.Field),
+		slog.String("type", string(err.Type)),
+		slog.String("message", err.Message),
+		slog.Any("details", err.err),
+	)
+}
+
+type marshallableError struct {
+	error
+}
+
+type marshalArgsError struct {
+	Field   string `json:"field"`
+	Type    string `json:"type,omitempty"`
+	Message string `json:"message"`
+	Details any    `json:"details,omitempty"`
+}
+
+func (m *marshallableError) MarshalJSON() ([]byte, error) {
+	return jsoniter.Marshal(m.Error())
+}
+
+func (err *ArgsError) MarshalJSON() ([]byte, error) {
+	var m = marshalArgsError{Field: err.Field, Type: err.Type, Message: err.Message}
+
+	if err.err == nil {
+		m.Details = nil
+		return jsoniter.Marshal(&m)
+	}
+
+	// check if the underlying error is a parse error
+	e := errors.Unwrap(err.err)
+	if e == nil {
+		e = err.err
+		m.Type = fmt.Sprintf("%T", e)
+	}
+
+	// need assertion because error doesn't know how to deal with marshalling
+	switch t := e.(type) {
+	case *types.ParseError,
+		*types.MinBoundsError,
+		*types.MaxBoundsError,
+		*types.RangeError,
+		*types.UnsupportedError:
+		m.Details = t
+	default:
+		m.Details = &marshallableError{e}
+	}
+
+	return jsoniter.Marshal(&m)
+}
+
+var (
+	maxBoundsErrorType   = fmt.Sprintf("%T", new(types.MaxBoundsError))
```

Comment:
Playing devils' advocate here: Why do we need all the custom error magic etc. to cover / validate queries in this manner, isn't that exactly what OpenAPI (which we already have in place) is for? If a caller (e.g. the Portal) wants to validate if the parameter is out of bounds or supported it simply has to run the query against the OpenAPI spec (without even having to perform a call) and be done with it. For that, the only thing required is that the OpenAPI spec is always in line with the internals (which it has to be anyway to be useful). And even if the validation should be provided as an API endpoint for some reason it would probably still be more simple to validate the query by running it against the OpenAPI spec internally.
I assume there are still some cases that are not covered by such mechanism, but at least all the ones I see here definitively are.
Given that these changes are quite substantial I'm just wondering if there is a reason why schema-based validation is out of the question?

Reply:
I think I understand your point only half way. The OpenAPI spec should validate the data structures returned.

Why I did it was to provide a means to also Unmarshal back into the data structures provided, e.g. for a golang client implementation.